plugins/optimization-detective/class-od-url-metric.php

```diff
@@ -65,6 +66,9 @@ final class OD_URL_Metric implements JsonSerializable {
 	 * @throws OD_Data_Validation_Exception When the input is invalid.
 	 */
 	public function __construct( array $data ) {
+		if ( ! array_key_exists( 'uuid', $data ) ) {
```

Comment:
`isset()` is faster than `array_key_exists()`, this has been subject of several core changes. I'd argue anyway that it would be safer to use the following here to ensure this is never empty:
```suggestion
		if ( ! isset( $data['uuid'] ) || ! $data['uuid'] ) {
```

Reply:
Sounds good, but what about https://github.com/WordPress/performance/issues/1219?

Reply:
I guess we should reduce the scope of that issue to remove use of `empty()` only.

Reply:
Also, the `! $data['uuid']` part is flagged by PHPStan's strict rules:

> Only booleans are allowed in a negated boolean, mixed given.

So I'll just commit the `! isset( $data['uuid'] )` part. The JSON Schema will handle validation of `uuid`. It should be absent for everyone right now, resulting in a new `uuid` being assigned to all existing URL metrics whenever a new URL metric is captured and added to an `od_url_metrics` post.

Done in 7a1b892f

Reply:
+1, I don't think there's any problem with `isset()` when used with array keys or object properties. In other words, `isset( $arr['something'] )` is perfectly fine, but `isset( $something )` is not.

It's a good idea to remove `empty()` usages and, if reasonable to implement, also warn about using `isset()` when it's not with array keys or object properties.

Reply:
> Only booleans are allowed in a negated boolean, mixed given.

On another note, this is the only thing that PHPStan complains about that I don't like. I don't like writing `'' === $value` just as little as I like writing `false === $value`, and I don't see any benefit of preventing `!` on a string. What _is_ the benefit? 🤔

Reply:
While I agree it can be annoying at times, I think the explicitness is good for type safety and making sure that you're always working with the data you expect. Otherwise, I had to ask Gemini for advice: https://g.co/gemini/share/270a3e76fcdd

@swissspidy may have more thoughts.